### ISSUE TITLE
Using function_type on a reference gives a non-obvious error message

### ISSUE THREAD

**Issue by ldionne** [badair/callable_traits#86]
Instead of failing with something like `function_type can't be used with ref-qualified types`, we get a SFINAE error saying that the library is unable to determine the parameters. This is somewhat misleading, and it took me a good couple of minutes to figure this one out.

**Comment by badair** [comment]
Ah, thanks.

Do you think the library should support reference types to function objects? We could simply remove reference at the top level, but at that point a case could be made to remove everything else as well.

**Comment by ldionne** [comment]
I don't know. I don't think it should, but we should at least get  a clear error message.

**Comment by badair** [comment]
@ldionne Do you think the "SFINAE errors" approach is more of a hindrance than a convenience?

**Comment by ldionne** [comment]
I don't know, but here's what I get:

```c++
#include <boost/callable_traits.hpp>
struct F { void operator()(int) {} };
using XX = boost::callable_traits::function_type<F&>::type;

int main() { }
```

Error:
```
boost/callable_traits/detail/sfinae_errors.hpp:34:9: error:
      no type named 'type' in
      'boost::callable_traits::error::parameters<boost::callable_traits::cannot_determine_parameters_for_this_type_>::_'
        using sfinae_try = typename CALLABLE_TRAITS_DISJUNCTION(
        ^~~~~
boost/callable_traits/detail/utility.hpp:48:1: note:
      in instantiation of template type alias 'sfinae_try' requested here
using fail_if_invalid = sfinae_try<T,
^
boost/callable_traits/function_type.hpp:24:1: note:
      in instantiation of template type alias 'fail_if_invalid' requested here
using function_type_t = //implementation-defined
^
boost/callable_traits/function_type.hpp:33:18: note:
      in instantiation of template type alias 'function_type_t' requested here
    using type = function_type_t<T>;
                 ^
test.cpp:7:36: note: in instantiation of template class
      'boost::callable_traits::function_type<F &>' requested here
using XX = boost::callable_traits::function_type<F&>::type;
                                   ^
1 error generated.
```

I thought somehow function objects were not supported or something weird like that. It could be the way it's worded: "unable to determine parameters" makes it sound like it's something you COULD determine, but don't because of some defficiency in the library.

**Comment by badair** [comment]
Pending CI testing, fixed via d87078752dd2769f444d1d2d531d6a500aae86d4.

SFINAE error message now reads:

error: no type named ‘type’ in ‘struct boost::callable_traits::error::reference_error<boost::callable_traits::reference_type_not_supported_by_this_metafunction_>::_’
